From 1d1e4c8717e1b83af31003ee1a9b309e7173a6b1 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Tue, 24 Jan 2023 19:47:26 +0100
Subject: [PATCH 271/841] feat(st-ddr): stm32mp2: implement standby software
 self-refresh

Distinguish stop and standby self-refresh entry functions for each mode.
Concerning software self-refresh, only IO retention enabling and DDC
(DRAM drift compensation, only for LPDDR4) disabling differ.
For exit function, use stm32mp_is_wakeup_from_standby() platform service
inside init sequence to execute relative actions. Add internal functions
to treat specific tasks, such as reset management.
Add a retention register in LPDDR4 list, in order to restore DDC value
after standby.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 800a2895994b3a8125582d7f84b43a835d2a6738 feat(st-ddr): stm32mp2: implement standby software self-refresh]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: If8e6673f3136ca1eb6b7e4d7121110bf127db6c4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/286975
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 .../phyinit/src/ddrphy_phyinit_reginterface.c |  2 +-
 .../ddrphy_phyinit_usercustom_saveretregs.c   |  7 +-
 drivers/st/ddr/stm32mp2_ddr.c                 | 89 ++++++++++++++---
 drivers/st/ddr/stm32mp2_ddr_helpers.c         | 99 ++++++++++++++++---
 drivers/st/ddr/stm32mp2_ram.c                 |  5 +-
 include/drivers/st/stm32mp2_ddr_helpers.h     |  3 +
 include/drivers/st/stm32mp2_ddr_regs.h        |  1 +
 7 files changed, 176 insertions(+), 30 deletions(-)

diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c
index 63cdcf78b..66755cf13 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c
@@ -31,7 +31,7 @@
 #if STM32MP_DDR3_TYPE || STM32MP_DDR4_TYPE
 #define MAX_NUM_RET_REGS	129
 #elif STM32MP_LPDDR4_TYPE
-#define MAX_NUM_RET_REGS	282
+#define MAX_NUM_RET_REGS	283
 #endif /* STM32MP_LPDDR4_TYPE */
 
 /*
diff --git a/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c b/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
index 0da09435b..17019707a 100644
--- a/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
+++ b/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021-2023, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -321,6 +321,11 @@ int ddrphy_phyinit_usercustom_saveretregs(void)
 		if (ret != 0) {
 			return ret;
 		}
+
+		ret = ddrphy_phyinit_trackreg(p_addr | TINITENG | CSR_SEQ0BDISABLEFLAG6_ADDR);
+		if (ret != 0) {
+			return ret;
+		}
 #endif /* STM32MP_LPDDR4_TYPE */
 	} /* p_addr */
 
diff --git a/drivers/st/ddr/stm32mp2_ddr.c b/drivers/st/ddr/stm32mp2_ddr.c
index 4b64ea6d2..7d386a971 100644
--- a/drivers/st/ddr/stm32mp2_ddr.c
+++ b/drivers/st/ddr/stm32mp2_ddr.c
@@ -242,6 +242,38 @@ static void ddr_reset(struct stm32mp_ddr_priv *priv)
 	udelay(DDR_DELAY_1US);
 }
 
+static void ddr_standby_reset(struct stm32mp_ddr_priv *priv)
+{
+	udelay(DDR_DELAY_1US);
+
+	mmio_write_32(priv->rcc + RCC_DDRCPCFGR,
+		      RCC_DDRCPCFGR_DDRCPEN | RCC_DDRCPCFGR_DDRCPLPEN | RCC_DDRCPCFGR_DDRCPRST);
+	mmio_write_32(priv->rcc + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRRST);
+	mmio_write_32(priv->rcc + RCC_DDRPHYCAPBCFGR,
+		      RCC_DDRPHYCAPBCFGR_DDRPHYCAPBEN | RCC_DDRPHYCAPBCFGR_DDRPHYCAPBLPEN |
+		      RCC_DDRPHYCAPBCFGR_DDRPHYCAPBRST);
+	mmio_write_32(priv->rcc + RCC_DDRCAPBCFGR,
+		      RCC_DDRCAPBCFGR_DDRCAPBEN | RCC_DDRCAPBCFGR_DDRCAPBLPEN |
+		      RCC_DDRCAPBCFGR_DDRCAPBRST);
+
+	mmio_clrbits_32(priv->rcc + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRPHYDLP);
+	mmio_setbits_32(priv->rcc + RCC_DDRPHYCCFGR, RCC_DDRPHYCCFGR_DDRPHYCEN);
+
+	udelay(DDR_DELAY_1US);
+}
+
+static void ddr_standby_reset_release(struct stm32mp_ddr_priv *priv)
+{
+	udelay(DDR_DELAY_1US);
+
+	mmio_write_32(priv->rcc + RCC_DDRCPCFGR, RCC_DDRCPCFGR_DDRCPEN | RCC_DDRCPCFGR_DDRCPLPEN);
+	mmio_write_32(priv->rcc + RCC_DDRITFCFGR, 0U);
+	mmio_clrbits_32(priv->rcc + RCC_DDRPHYCAPBCFGR, RCC_DDRPHYCAPBCFGR_DDRPHYCAPBRST);
+	mmio_write_32(priv->rcc + RCC_DDRCFGR, RCC_DDRCFGR_DDRCFGEN | RCC_DDRCFGR_DDRCFGLPEN);
+
+	udelay(DDR_DELAY_1US);
+}
+
 static void ddr_sysconf_configuration(struct stm32mp_ddr_priv *priv)
 {
 	mmio_write_32(stm32_ddrdbg_get_base() + DDRDBG_LP_DISABLE,
@@ -349,8 +381,7 @@ void stm32mp2_ddr_init(struct stm32mp_ddr_priv *priv,
 	}
 
 	/* Check DDR PHY pads retention */
-	ddr_retdis = mmio_read_32((uint32_t)(priv->pwr) + PWR_CR11) &
-		     PWR_CR11_DDRRETDIS;
+	ddr_retdis = mmio_read_32(priv->pwr + PWR_CR11) & PWR_CR11_DDRRETDIS;
 	if (config->self_refresh) {
 		if (ddr_retdis == PWR_CR11_DDRRETDIS) {
 			VERBOSE("self-refresh aborted: no retention\n");
@@ -358,32 +389,51 @@ void stm32mp2_ddr_init(struct stm32mp_ddr_priv *priv,
 		}
 	}
 
-	if (!config->self_refresh) {
+	if (config->self_refresh) {
+		ddr_standby_reset(priv);
+
 		VERBOSE("disable DDR PHY retention\n");
-		mmio_setbits_32((uint32_t)(priv->pwr) + PWR_CR11, PWR_CR11_DDRRETDIS);
-	}
+		mmio_setbits_32(priv->pwr + PWR_CR11, PWR_CR11_DDRRETDIS);
+
+		udelay(DDR_DELAY_1US);
+
+		mmio_clrbits_32(priv->rcc + RCC_DDRCAPBCFGR, RCC_DDRCAPBCFGR_DDRCAPBRST);
 
-	ddr_reset(priv);
+		udelay(DDR_DELAY_1US);
+
+	} else {
+		VERBOSE("disable DDR PHY retention\n");
+		mmio_setbits_32(priv->pwr + PWR_CR11, PWR_CR11_DDRRETDIS);
+
+		ddr_reset(priv);
 
-	ddr_sysconf_configuration(priv);
+		ddr_sysconf_configuration(priv);
+	}
 
 	stm32mp_ddr_set_reg(priv, REG_REG, &config->c_reg, ddr_registers);
 	stm32mp_ddr_set_reg(priv, REG_TIMING, &config->c_timing, ddr_registers);
 	stm32mp_ddr_set_reg(priv, REG_MAP, &config->c_map, ddr_registers);
 	stm32mp_ddr_set_reg(priv, REG_PERF, &config->c_perf, ddr_registers);
 
-	/*  DDR core and PHY reset de-assert */
-	mmio_clrbits_32(priv->rcc + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRRST);
+	if (!config->self_refresh) {
+		/*  DDR core and PHY reset de-assert */
+		mmio_clrbits_32(priv->rcc + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRRST);
+
+		disable_refresh(priv->ctl);
+	}
 
-	disable_refresh(priv->ctl);
+	if (config->self_refresh) {
+		ddr_standby_reset_release(priv);
 
-	if (false /* TODO stm32mp1_is_wakeup_from_standby() */) {
 		/* Initialize DDR by skipping training and disabling result saving */
 		ret = ddrphy_phyinit_sequence(true, false);
 
 		if (ret == 0) {
 			ret = ddrphy_phyinit_restore_sequence();
 		}
+
+		/* Poll on ddrphy_initeng0_phyinlpx.phyinlp3 = 0 */
+		ddr_wait_lp3_mode(false);
 	} else {
 		/* Initialize DDR including training and result saving */
 		ret = ddrphy_phyinit_sequence(false, true);
@@ -396,7 +446,22 @@ void stm32mp2_ddr_init(struct stm32mp_ddr_priv *priv,
 
 	ddr_activate_controller(priv->ctl, false);
 
-	restore_refresh(priv->ctl, config->c_reg.rfshctl3, config->c_reg.pwrctl);
+	if (config->self_refresh) {
+		struct stm32mp_ddrctl *ctl = priv->ctl;
+
+		/* SW self refresh exit prequested */
+		mmio_clrbits_32((uintptr_t)&ctl->pwrctl, DDRCTRL_PWRCTL_SELFREF_SW);
+
+		if (ddr_sr_exit_loop() != 0) {
+			ERROR("DDR Standby exit error\n");
+			panic();
+		}
+
+		/* Re-enable DFI low-power interface */
+		mmio_clrbits_32((uintptr_t)&ctl->dfilpcfg0, DDRCTRL_DFILPCFG0_DFI_LP_EN_SR);
+	} else {
+		restore_refresh(priv->ctl, config->c_reg.rfshctl3, config->c_reg.pwrctl);
+	}
 
 	stm32mp_ddr_enable_axi_port(priv->ctl);
 }
diff --git a/drivers/st/ddr/stm32mp2_ddr_helpers.c b/drivers/st/ddr/stm32mp2_ddr_helpers.c
index 1062e8260..1665d08d3 100644
--- a/drivers/st/ddr/stm32mp2_ddr_helpers.c
+++ b/drivers/st/ddr/stm32mp2_ddr_helpers.c
@@ -118,7 +118,27 @@ void ddr_activate_controller(struct stm32mp_ddrctl *ctl, bool sr_entry)
 	unset_qd1_qd3_update_conditions(ctl);
 }
 
-static void wait_lp3_mode(bool sr_entry)
+#if STM32MP_LPDDR4_TYPE
+static void disable_phy_ddc(void)
+{
+	/* Enable APB access to internal CSR registers */
+	mmio_write_32(stm32mp_ddrphyc_base() + DDRPHY_APBONLY0_MICROCONTMUXSEL, 0);
+	mmio_write_32(stm32mp_ddrphyc_base() + DDRPHY_DRTUB0_UCCLKHCLKENABLES,
+		      DDRPHY_DRTUB0_UCCLKHCLKENABLES_UCCLKEN |
+		      DDRPHY_DRTUB0_UCCLKHCLKENABLES_HCLKEN);
+
+	/* Disable DRAM drift compensation */
+	mmio_write_32(stm32mp_ddrphyc_base() + DDRPHY_INITENG0_P0_SEQ0BDISABLEFLAG6, 0xFFFF);
+
+	/* Disable APB access to internal CSR registers */
+	mmio_write_32(stm32mp_ddrphyc_base() + DDRPHY_DRTUB0_UCCLKHCLKENABLES,
+		      DDRPHY_DRTUB0_UCCLKHCLKENABLES_HCLKEN);
+	mmio_write_32(stm32mp_ddrphyc_base() + DDRPHY_APBONLY0_MICROCONTMUXSEL,
+		      DDRPHY_APBONLY0_MICROCONTMUXSEL_MICROCONTMUXSEL);
+}
+#endif /* STM32MP_LPDDR4_TYPE */
+
+void ddr_wait_lp3_mode(bool sr_entry)
 {
 	uint64_t timeout;
 	bool repeat_loop = false;
@@ -202,7 +222,7 @@ static int sr_entry_loop(void)
 	return sr_loop(true);
 }
 
-static int sr_exit_loop(void)
+int ddr_sr_exit_loop(void)
 {
 	return sr_loop(false);
 }
@@ -235,6 +255,13 @@ static int ssr_entry(bool standby)
 		panic();
 	}
 
+#if STM32MP_LPDDR4_TYPE
+	if (standby) {
+		/* Disable DRAM drift compensation */
+		disable_phy_ddc();
+	}
+#endif /* STM32MP_LPDDR4_TYPE */
+
 	disable_dfi_low_power_interface((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	/* SW self refresh entry prequested */
@@ -250,7 +277,7 @@ static int ssr_entry(bool standby)
 	ddr_activate_controller((struct stm32mp_ddrctl *)ddrctrl_base, true);
 
 	/* Poll on ddrphy_initeng0_phyinlpx.phyinlp3 = 1 */
-	wait_lp3_mode(true);
+	ddr_wait_lp3_mode(true);
 
 	if (standby) {
 		mmio_clrbits_32(stm32mp_pwr_base() + PWR_CR11, PWR_CR11_DDRRETDIS);
@@ -269,6 +296,11 @@ static int sr_ssr_entry(void)
 	return ssr_entry(false);
 }
 
+static int stdby_sr_ssr_entry(void)
+{
+	return ssr_entry(true);
+}
+
 static int sr_ssr_exit(void)
 {
 	uintptr_t ddrctrl_base = stm32mp_ddrctrl_base();
@@ -284,12 +316,12 @@ static int sr_ssr_exit(void)
 	ddr_activate_controller((struct stm32mp_ddrctl *)ddrctrl_base, false);
 
 	/* Poll on ddrphy_initeng0_phyinlpx.phyinlp3 = 0 */
-	wait_lp3_mode(false);
+	ddr_wait_lp3_mode(false);
 
 	/* SW self refresh exit prequested */
 	mmio_clrbits_32(ddrctrl_base + DDRCTRL_PWRCTL, DDRCTRL_PWRCTL_SELFREF_SW);
 
-	if (sr_exit_loop() != 0) {
+	if (ddr_sr_exit_loop() != 0) {
 		return -1;
 	}
 
@@ -326,19 +358,29 @@ static int sr_hsr_set(void)
 	return 0;
 }
 
-static int sr_hsr_entry(void)
+static int hsr_entry(bool standby __unused)
 {
 	mmio_write_32(stm32mp_rcc_base() + RCC_DDRCPCFGR, RCC_DDRCPCFGR_DDRCPLPEN);
 
 	return sr_entry_loop(); /* read_data should be equal to 0x223 */
 }
 
+static int sr_hsr_entry(void)
+{
+	return hsr_entry(false);
+}
+
+static int stdby_sr_hsr_entry(void)
+{
+	return hsr_entry(true);
+}
+
 static int sr_hsr_exit(void)
 {
 	mmio_write_32(stm32mp_rcc_base() + RCC_DDRCPCFGR,
 		      RCC_DDRCPCFGR_DDRCPLPEN | RCC_DDRCPCFGR_DDRCPEN);
 
-	/* TODO: check if sr_exit_loop() is needed here */
+	/* TODO: check if ddr_sr_exit_loop() is needed here */
 
 	return 0;
 }
@@ -350,7 +392,7 @@ static int sr_asr_set(void)
 	return 0;
 }
 
-static int sr_asr_entry(void)
+static int asr_entry(bool standby __unused)
 {
 	/*
 	 * Automatically enter into self refresh when there is no ddr traffic
@@ -360,14 +402,24 @@ static int sr_asr_entry(void)
 	return sr_entry_loop();
 }
 
+static int sr_asr_entry(void)
+{
+	return asr_entry(false);
+}
+
+static int stdby_sr_asr_entry(void)
+{
+	return asr_entry(true);
+}
+
 static int sr_asr_exit(void)
 {
-	return sr_exit_loop();
+	return ddr_sr_exit_loop();
 }
 
 int ddr_sw_self_refresh_exit(void)
 {
-	int ret = -1;
+	int ret = -EINVAL;
 
 	switch (saved_ddr_sr_mode) {
 	case DDR_SSR_MODE:
@@ -393,9 +445,9 @@ uint32_t ddr_get_io_calibration_val(void)
 	return 0U;
 }
 
-int ddr_standby_sr_entry(void)
+int ddr_sr_entry(void)
 {
-	int ret = -1;
+	int ret = -EINVAL;
 
 	switch (saved_ddr_sr_mode) {
 	case DDR_SSR_MODE:
@@ -414,6 +466,27 @@ int ddr_standby_sr_entry(void)
 	return ret;
 }
 
+int ddr_standby_sr_entry(void)
+{
+	int ret = -EINVAL;
+
+	switch (saved_ddr_sr_mode) {
+	case DDR_SSR_MODE:
+		ret = stdby_sr_ssr_entry();
+		break;
+	case DDR_HSR_MODE:
+		ret = stdby_sr_hsr_entry();
+		break;
+	case DDR_ASR_MODE:
+		ret = stdby_sr_asr_entry();
+		break;
+	default:
+		break;
+	}
+
+	return ret;
+}
+
 enum stm32mp2_ddr_sr_mode ddr_read_sr_mode(void)
 {
 	uint32_t pwrctl = mmio_read_32(stm32mp_ddrctrl_base() + DDRCTRL_PWRCTL);
@@ -439,7 +512,7 @@ enum stm32mp2_ddr_sr_mode ddr_read_sr_mode(void)
 
 void ddr_set_sr_mode(enum stm32mp2_ddr_sr_mode mode)
 {
-	int ret = -1;
+	int ret = -EINVAL;
 
 	if (mode == saved_ddr_sr_mode) {
 		return;
diff --git a/drivers/st/ddr/stm32mp2_ram.c b/drivers/st/ddr/stm32mp2_ram.c
index 045ae5aae..672331fa6 100644
--- a/drivers/st/ddr/stm32mp2_ram.c
+++ b/drivers/st/ddr/stm32mp2_ram.c
@@ -126,10 +126,9 @@ static int stm32mp2_ddr_setup(void)
 
 	config.self_refresh = false;
 
-	/* TODO if (stm32mp1_is_wakeup_from_standby()) {
+	if (stm32mp_is_wakeup_from_standby()) {
 		config.self_refresh = true;
-		config.zdata = stm32_get_zdata_from_context();
-	}*/
+	}
 
 	/*  Map dynamically RETRAM area to save or restore PHY retention registers */
 	if (stm32mp_map_retram() != 0) {
diff --git a/include/drivers/st/stm32mp2_ddr_helpers.h b/include/drivers/st/stm32mp2_ddr_helpers.h
index 2f667390e..984f0cc05 100644
--- a/include/drivers/st/stm32mp2_ddr_helpers.h
+++ b/include/drivers/st/stm32mp2_ddr_helpers.h
@@ -20,8 +20,11 @@ enum stm32mp2_ddr_sr_mode {
 };
 
 void ddr_activate_controller(struct stm32mp_ddrctl *ctl, bool sr_entry);
+void ddr_wait_lp3_mode(bool state);
+int ddr_sr_exit_loop(void);
 int ddr_sw_self_refresh_exit(void);
 uint32_t ddr_get_io_calibration_val(void);
+int ddr_sr_entry(void);
 int ddr_standby_sr_entry(void);
 enum stm32mp2_ddr_sr_mode ddr_read_sr_mode(void);
 void ddr_set_sr_mode(enum stm32mp2_ddr_sr_mode mode);
diff --git a/include/drivers/st/stm32mp2_ddr_regs.h b/include/drivers/st/stm32mp2_ddr_regs.h
index bb0271df7..78a3e9003 100644
--- a/include/drivers/st/stm32mp2_ddr_regs.h
+++ b/include/drivers/st/stm32mp2_ddr_regs.h
@@ -16,6 +16,7 @@ struct stm32mp_ddrphy {
 } __packed;
 
 /* DDRPHY registers offsets */
+#define DDRPHY_INITENG0_P0_SEQ0BDISABLEFLAG6	U(0x240004)
 #define DDRPHY_INITENG0_P0_PHYINLPX		U(0x2400A0)
 #define DDRPHY_DRTUB0_UCCLKHCLKENABLES		U(0x300200)
 #define DDRPHY_APBONLY0_MICROCONTMUXSEL		U(0x340000)
-- 
2.47.2

